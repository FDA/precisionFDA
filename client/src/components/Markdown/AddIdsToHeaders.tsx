import React, { useEffect } from 'react'

export function AddIdsToHeaders({ docRef, content }: { docRef: any, content: string }) {
  useEffect(() => {
    if (docRef.current) {
      const headings = docRef.current.querySelectorAll('h1, h2, h3')

      headings.forEach((heading: any) => {
        // eslint-disable-next-line no-param-reassign
        heading.id = heading.getAttribute('id') || heading.innerText
          .toLowerCase()
          // eslint-disable-next-line no-useless-escape
          .replace(/[`~!@#$%^&*()_|+\-=?;:'",.<>\{\}\[\]\\\/]/gi, '')
          .replace(/ +/g, '-')
      })
    }
  }, [docRef])
  
  return (
    <div ref={docRef} dangerouslySetInnerHTML={{ __html: content }} />
  )
}
