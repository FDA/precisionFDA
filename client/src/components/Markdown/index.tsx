import React, { Ref } from 'react'
import { ReactMarkdown } from 'react-markdown/lib/react-markdown'
import rehypeHighlight from 'rehype-highlight/lib'
import rehypeRaw from 'rehype-raw'
import rehypeSlug from 'rehype-slug'
import remarkGfm from 'remark-gfm'
import styled from 'styled-components'

export const MarkdownStyle = styled.div`
  padding: 16px;
  max-width: 800px;
  line-height: 24px;

  img {
    object-fit: cover;
    object-position: center;
    max-width: 100%;
  }

  p {
    margin-bottom: 24px;
  }

  h1 {
    margin-top: 3rem;
    margin-bottom: 1rem;
    font-size: 32px;
    color: #333333;
  }
  h2, h3, h4 {
    margin-top: 2.5rem;
    margin-bottom: 1rem;
  }

  code {
    padding: 2px 4px;
    font-size: 90%;
    color: #c7254e;
    background-color: #f9f2f4;
    border-radius: 3px;
  }

  pre {
    code {
      -webkit-text-size-adjust: 100%;
      -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
      box-sizing: border-box;
      overflow: auto;
      display: block;
      padding: 9.5px;
      margin: 0 0 10px;
      font-size: 13px;
      line-height: 1.428571429;
      word-break: break-all;
      word-wrap: break-word;
      color: #333333;
      background-color: #f5f5f5;
      border: 1px solid #ccc;
      border-radius: 3px;
      white-space: pre-line;
    }

  }
`

export const Markdown = ({ docRef, data = '' }: { docRef: Ref<any>, data: string }) => {
  return (
    <div ref={docRef}>
      <ReactMarkdown
        remarkPlugins={[remarkGfm]}
        rehypePlugins={[rehypeSlug, rehypeHighlight, rehypeRaw]}
        >
        {data}
      </ReactMarkdown>
    </div>
  )
}
