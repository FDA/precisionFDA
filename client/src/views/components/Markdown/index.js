import React from 'react'
import { Remarkable } from 'remarkable'
import { linkify } from 'remarkable/linkify'
import PropTypes from 'prop-types'


const Markdown = ({ data = '', ...rest }) => {
  const md = new Remarkable('full', {
    typographer: true,
  }).use(linkify)

  return <div dangerouslySetInnerHTML={{ __html: md.render(data) }} {...rest}></div>
}

Markdown.propTypes = {
  data: PropTypes.string,
}

export default Markdown
