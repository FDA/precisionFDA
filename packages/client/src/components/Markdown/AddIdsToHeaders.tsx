import React, { useEffect } from 'react'
import styled from 'styled-components'

const Content = styled.div``

export function AddIdsToHeaders({ docRef, content, as }: { docRef: any, content: string, as: any }) {
  useEffect(() => {
    if (docRef.current) {
      const headings = docRef.current.querySelectorAll('h1, h2, h3')

      headings.forEach((heading: any) => {
        // eslint-disable-next-line no-param-reassign
        heading.id = heading.getAttribute('id') || heading.innerText
          .toLowerCase()
          // eslint-disable-next-line no-useless-escape
          .replace(/[`~!@#$%^&*()_|+\-=?;:'",.<>\{\}\[\]\\\/]/gi, '')
          .replace(/ +/g, '-')
      })
    }
  }, [docRef])
  
  return (
    <Content as={as} ref={docRef} dangerouslySetInnerHTML={{ __html: content }} />
  )
}
