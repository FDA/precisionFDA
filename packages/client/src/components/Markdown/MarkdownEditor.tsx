/* eslint-disable react/jsx-no-useless-fragment */
import React from 'react'
import styled from 'styled-components'
import { ControllerRenderProps } from 'react-hook-form'
import { Markdown } from '.'

import { ITab, TabsSwitch } from '../TabsSwitch'
import { StyledMarkdown } from '../../features/discussions/styles'
import { MarkdownIcon } from '../icons/MarkdownIcon'
import ExternalLink from '../Controls/ExternalLink'

export const Help = styled.div`
  padding: 6px;
  padding-right: 12px;
  background-color: var(--tertiary-100);
  border-left: 5px solid var(--c-layout-border);
  color: var(--c-text-400);
  text-align: left;
  font-size: 13px;
  width: fit-content;
`

export const WeMarkdown = styled.div`
  width: fit-content;
  display: flex;
  align-items: center;
  gap: 4px;
  font-size: 12px;
  color: var(--c-text-500);
  opacity: 0.6;
  transition: opacity 0.08s linear 0.08s;

  svg {
    color: var(--primary-600);
  }
  &:hover {
      opacity: 1;
  }
`
export const StyledMarkdownHelper = styled.div`
  width: fit-content;
`

export const StyedEditor = styled.div`
  padding: 8px;

  --spacing: 8px;
  textarea {
    min-width: calc(100% - var(--spacing) * 2);
    max-width: calc(100% - var(--spacing) * 2);
    padding: var(--spacing);
    min-height: 70px;
    border-radius: 5px;
    border: 1px solid var(--tertiary-250);
    background: var(--c-textarea-bg);
    height: 20vh;

    &:focus {
      border-color: var(--primary-600);
      border-right-width: 1px !important;
      outline: 0;
      box-shadow: 0 0 0 2px rgba(24, 144, 255, 0.2);
    }
  }
`
const StyledPreview = styled(StyledMarkdown)`
  padding: 8px 16px;
  min-height: 74px;
`

export const MarkdownEditor = ({
  disabled,
  field,
}: {
  disabled: boolean
  field: ControllerRenderProps
}) => {
  return (
    <TabsSwitch
      transparent
      tabsConfig={
        [
          {
            header: 'Edit',
            tab: (
              <StyedEditor>
                <textarea
                  placeholder="Leave a comment"
                  disabled={disabled}
                  {...field}
                />
                <StyledMarkdownHelper><ExternalLink to="https://www.markdownguide.org/basic-syntax/"><WeMarkdown><MarkdownIcon height={17} />Markdown Support</WeMarkdown></ExternalLink></StyledMarkdownHelper>
              </StyedEditor>
            ),
            hide: false,
          },
          {
            header: 'Preview',
            tab: (
              <StyledPreview>
                <Markdown data={field.value ?? ''} docRef={null} />
              </StyledPreview>
            ),
            hide: false,
          },
        ] satisfies ITab[]
      }
    />
  )
}
