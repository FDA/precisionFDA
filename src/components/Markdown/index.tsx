import React from 'react'
import { Remarkable } from 'remarkable'
import { linkify } from 'remarkable/linkify'
import styled from 'styled-components'

const StyledMarkdown = styled.div`
  padding: 16px;
`

export const Markdown = ({ data = '', ...rest }: { data: string }) => {
  const md = new Remarkable('full', {
    typographer: true,
  }).use(linkify)

  return <StyledMarkdown dangerouslySetInnerHTML={{ __html: md.render(data) }} {...rest}></StyledMarkdown>
}
