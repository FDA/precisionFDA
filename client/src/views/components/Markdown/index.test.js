import React from 'react'
import { shallow } from 'enzyme'

import Markdown from './index'


describe('Markdown', () => {
  it('matches snapshot', () => {
    const data = '# h1 Heading'
    const component = shallow(<Markdown data={data} />)
    expect(component).toMatchSnapshot()
  })
})
