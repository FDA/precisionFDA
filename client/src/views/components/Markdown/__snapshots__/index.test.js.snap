// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown matches snapshot 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<h1>h1 Heading</h1>
",
    }
  }
/>
`;
